packages/lib/src/load_plugin_list.ts

```diff
@@ -0,0 +1,184 @@
+import fs from "fs-extra";
+import path from "path";
+import { logger } from "./logging";
+import * as libFileOps from "./file_ops";
+import vm from "vm";
+
+/**
+ * Searches for local plugins in the plugins/ and external_plugins/ directories
+ * and adds them to the plugin list if they contain a package.json file.
+ *
+ * @param pluginList - Map of plugin names to their file paths
+ * @param pluginListPath - Path to the JSON file storing the plugin list
+ * @returns Promise that resolves when plugin discovery is complete
+ */
+async function findLocalPlugins(pluginList: Map<string, string>, pluginListPath: string) {
+	// Check folders for plugins
+	const pluginFolders = [
+		"plugins",
+		"external_plugins",
+	];
+
+	let has_changed = 0;
+	const ignoredFolders = new Set(["node_modules", ".git", "dist", "build", "disabled"]);
+	const maxDepth = 3;
+
+	for (const pluginFolder of pluginFolders) {
+		let pluginNames;
+		try {
+			pluginNames = await fs.readdir(pluginFolder);
+		} catch (err: any) {
+			if (err.code === "ENOENT") {
+				// Skip if folder doesn't exist
+				continue;
+			}
+			throw err;
+		}
+		for (const pluginName of pluginNames) {
+			const pluginPath = path.join(pluginFolder, pluginName);
+			if (pluginList.has(pluginName)) {
+				// If the one in the list is not the same as this one, warn
+				if (pluginList.get(pluginName) !== path.resolve(pluginPath)) {
+					logger.warn(
+						`${pluginName} is already in the plugin list, but with a different path - ` +
+						`using ${pluginList.get(pluginName)}`
+					);
+				}
+				continue;
+			}
+			const stats = await fs.stat(pluginPath);
+			if (stats.isDirectory()) {
+				// Check for package.json and plugin keyword
+				if (await checkPackageJson(pluginPath)) {
+					pluginList.set(pluginName, path.resolve(pluginPath));
+					logger.info(`Added ${pluginName} from ${pluginFolder}`);
+					has_changed += 1;
+					continue;
+				}
+
+				if (pluginPath.split(path.sep).length < maxDepth && !ignoredFolders.has(pluginName)) {
+					logger.warn(`Possible monorepo detected, scanning ${pluginPath} for clusterio-plugins`);
```

Comment:
I think warn makes sense - this codepath gets triggered for incorrectly set up plugins. If you don't pay attention it can be difficult to figure out why your plugin isn't being loaded. The warn allows you to question yourself on whether you are developing in a monorepo or not.

It will never warn in production since people don't use monorepos in production (I think)

Reply:
Good point on monorepos not existing in production, they will instead be added via npm. I am happy to have this remain as a warning.